### PR TITLE
feat(redis): add redis implementation of `KvBufferStore`

### PR DESCRIPTION
**Pull request by dtfiedler**
This implementation is farily straightforward. We will add a `redis` container to the `docker-compose` and set is as the default `KvBufferStore` for header and transaction cache data once we validate it works as expected.

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -54,6 +54,16 @@ services:
...
+      - REDIS_CACHE_URL=${REDIS_CACHE_URL:-redis://redis:6379}
+    depends_on:
+      - redis
+
+  redis:
+    image: redis:latest
+    ports:
+      - "6379:6379"
+    volumes:
+      - ${REDIS_DATA_PATH:-./data/redis}:/app/data/redis
```
Comment (dtfiedler) [inline docker-compose.yaml:66]: i believe we need to provide either some environment variables or a `redis.conf` so redis saves to disk on certain intervals

File: docker-compose.yaml [hunk 2]
```diff
@@ -53,7 +53,19 @@ services:
       - SANDBOX_PROTOCOL=${SANDBOX_PROTOCOL:-}
       - START_WRITERS=${START_WRITERS:-}
       - CONTRACT_ID=${CONTRACT_ID:-}
-      - CHAIN_CACHE_TYPE=${CHAIN_CACHE_TYPE:-}
+      - CHAIN_CACHE_TYPE=${CHAIN_CACHE_TYPE:-redis}
+      - REDIS_CACHE_URL=${REDIS_CACHE_URL:-redis://redis:6379}
+    depends_on:
+      - redis
```
Comment (dtfiedler) [inline docker-compose.yaml:59]: we don't use this for other services, so may remove

File: src/store/redis-kv-store.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import { RedisClientType, commandOptions, createClient } from 'redis';
+import winston from 'winston';
+
+import { KVBufferStore } from '../types.js';
+
+export class RedisKvStore implements KVBufferStore {
+  private client: RedisClientType;
+  private log: winston.Logger;
+
+  constructor({ log, redisUrl }: { log: winston.Logger; redisUrl: string }) {
```
[the remark this answers is not in the document]
Comment (dtfiedler) [reply src/store/redis-kv-store.ts]: [df0e4f4](https://github.com/ar-io/ar-io-node/pull/62/commits/df0e4f40260729e461887871dc90c88114c9925e)

File: src/store/redis-kv-store.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    this.log = log.child({ class: this.constructor.name });
+    this.client = createClient({
+      url: redisUrl,
+    });
+    this.client.on('error', (err) => {
+      this.log.error(`Redis error: ${err}`);
```
[the remark this answers is not in the document]
Comment (dtfiedler) [reply src/store/redis-kv-store.ts:43]: will add - do you think it's worth adding metrics for cache hits/misses?

[the remark this answers is not in the document]
Comment (djwhitt) [reply src/store/redis-kv-store.ts:44]: Initial connection error is good to have, but not exactly what I was intending. I was wanting one in the `on('error'...` handler. Still fine to keep the connect error counter too.

File: docker-compose.yaml [hunk 5]
```diff
@@ -53,7 +56,20 @@ services:
...
+      - CHAIN_CACHE_TYPE=${CHAIN_CACHE_TYPE:-redis}
+      - REDIS_CACHE_URL=${REDIS_CACHE_URL:-redis://redis:6379}
+      - REDIS_CACHE_TTL_SECONDS=${REDIS_CACHE_TTL_SECONDS:-r}
+    depends_on:
+      - redis
+
+  redis:
+    image: redis:latest
+    # TODO: append only is reccomended to prevent losing data
+    # command: redis-server --appendonly yes
```
Comment (djwhitt) [inline docker-compose.yaml:68]: Setting `appendonly` sounds good. We should also limit max memory to something reasonable (2GB seems fine).

**Review by djwhitt — COMMENTED** [review]
Added a couple comments about the metrics and Redis configuration. After those are resolved I think this will be ready to merge.

File: src/store/redis-kv-store.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtfiedler) [reply src/store/redis-kv-store.ts:44]: apologies - complete misread on my part. added in [2dcf902](https://github.com/ar-io/ar-io-node/pull/62/commits/2dcf902c0d9b9583ca72125fd2ff8b86d3063604)

File: docker-compose.yaml (hunk 5 above)
Comment (dtfiedler) [reply docker-compose.yaml]: updated here and allowed an env var should operators have a preference [2dcf902](https://github.com/ar-io/ar-io-node/pull/62/commits/2dcf902c0d9b9583ca72125fd2ff8b86d3063604)

**Review by djwhitt — APPROVED** [review]
Looks good. Nice work!

**Pull request merged**